ext-apiserver/pkg/controller/machine/node.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+import (
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// We are using "machine" annotation to link node and machine resource. The "machine"
+// annotation is an implementation detail of how the two objects can get linked together, but it is
+// not required behavior. However, in the event that a Machine.Spec update requires replacing the
+// Node, this can allow for faster turn-around time by allowing a new Node to be created with a new
+// name while the old node is being deleted.
+//
+// Currently, these annotations are added by the node itself as part of its
+// bootup script after "kubeadm join" succeeds.
+func (c *MachineControllerImpl) link(node *corev1.Node) {
+	if val, _ := c.linkedNodes[node.ObjectMeta.Name]; val {
+		return
+	}
+
+	if val, ok := node.ObjectMeta.Annotations["machine"]; ok {
+		machine, err := c.machineClient.Get(val, metav1.GetOptions{})
+		if err != nil {
+			glog.Errorf("Error getting machine %v: %v\n", val, err)
+			return
+		}
+
+		machine.Status.NodeRef = objectRef(node)
+
+		if _, err := c.machineClient.UpdateStatus(machine); err != nil {
```

Comment:
Should there be some retry logic on this failure? Will this update be requeued if it fails?

Reply:
Good point. Immediately retry won't much sense. If API server is not in a good condition, it will take it some time to recover. The same issue is also for unlink. It's less important for unlink though as machine is either going away or being unlinked. I will do some investigation to how we can do retry here.